In [72]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn import metrics
from sklearn.linear_model import LinearRegression

# Read in the datasets:

In [73]:
train_df = pd.read_csv("../datasets/train.csv")

In [74]:
test_df = pd.read_csv("../datasets/test.csv")

# Investigate the data:

### "Train" dataset:

In [75]:
train_df.shape

(2051, 81)

In [76]:
train_df.head(15)

Id        PID  MS SubClass MS Zoning  Lot Frontage  Lot Area Street  \
0    109  533352170           60        RL           NaN     13517   Pave   
1    544  531379050           60        RL          43.0     11492   Pave   
2    153  535304180           20        RL          68.0      7922   Pave   
3    318  916386060           60        RL          73.0      9802   Pave   
4    255  906425045           50        RL          82.0     14235   Pave   
5    138  535126040           20        RL         137.0     16492   Pave   
6   2827  908186070          180        RM          35.0      3675   Pave   
7    145  535154050           20        RL           NaN     12160   Pave   
8   1942  535353130           20        RL           NaN     15783   Pave   
9   1956  535426130           60        RL          70.0     11606   Pave   
10  1044  527451290          160        RM          21.0      1680   Pave   
11  2752  906380150           20        RL          64.0      7488   Pave   
12   807  906226060           70        RL         120.0     26400   Pave   
13  1177  533236070          160        FV          24.0      2645   Pave   
14  1480  907418040           20        RL          74.0      9627   Pave   

   Alley Lot Shape Land Contour  ... Screen Porch Pool Area Pool QC  Fence  \
0    NaN       IR1          Lvl  ...            0         0     NaN    NaN   
1    NaN       IR1          Lvl  ...            0         0     NaN    NaN   
2    NaN       Reg          Lvl  ...            0         0     NaN    NaN   
3    NaN       Reg          Lvl  ...            0         0     NaN    NaN   
4    NaN       IR1          Lvl  ...            0         0     NaN    NaN   
5    NaN       IR1          Lvl  ...            0         0     NaN    NaN   
6    NaN       Reg          Lvl  ...            0         0     NaN    NaN   
7    NaN       IR1          Lvl  ...            0         0     NaN  MnPrv   
8    NaN       Reg          Lvl  ...            0         0     NaN  MnPrv   
9    NaN       IR1          HLS  ...            0         0     NaN    NaN   
10   NaN       Reg          Lvl  ...            0         0     NaN    NaN   
11   NaN       IR1          Lvl  ...            0         0     NaN    NaN   
12   NaN       Reg          Bnk  ...            0         0     NaN    NaN   
13  Pave       Reg          Lvl  ...            0         0     NaN    NaN   
14   NaN       Reg          Lvl  ...            0         0     NaN    NaN   

   Misc Feature Misc Val Mo Sold Yr Sold  Sale Type  SalePrice  
0           NaN        0       3    2010        WD      130500  
1           NaN        0       4    2009        WD      220000  
2           NaN        0       1    2010        WD      109000  
3           NaN        0       4    2010        WD      174000  
4           NaN        0       3    2010        WD      138500  
5           NaN        0       6    2010        WD      190000  
6           NaN        0       6    2006        New     140000  
7           NaN        0       5    2010        COD     142000  
8          Shed      400       6    2007        WD      112500  
9           NaN        0       9    2007        WD      135000  
10          NaN        0       7    2008        COD      85400  
11          NaN        0       2    2006        WD      183600  
12          NaN        0       6    2009        WD      131000  
13          NaN        0      12    2008      ConLD     200000  
14          NaN        0       6    2008        WD      193000  

[15 rows x 81 columns]

In [77]:
train_df.columns

Index(['Id', 'PID', 'MS SubClass', 'MS Zoning', 'Lot Frontage', 'Lot Area',
       'Street', 'Alley', 'Lot Shape', 'Land Contour', 'Utilities',
       'Lot Config', 'Land Slope', 'Neighborhood', 'Condition 1',
       'Condition 2', 'Bldg Type', 'House Style', 'Overall Qual',
       'Overall Cond', 'Year Built', 'Year Remod/Add', 'Roof Style',
       'Roof Matl', 'Exterior 1st', 'Exterior 2nd', 'Mas Vnr Type',
       'Mas Vnr Area', 'Exter Qual', 'Exter Cond', 'Foundation', 'Bsmt Qual',
       'Bsmt Cond', 'Bsmt Exposure', 'BsmtFin Type 1', 'BsmtFin SF 1',
       'BsmtFin Type 2', 'BsmtFin SF 2', 'Bsmt Unf SF', 'Total Bsmt SF',
       'Heating', 'Heating QC', 'Central Air', 'Electrical', '1st Flr SF',
       '2nd Flr SF', 'Low Qual Fin SF', 'Gr Liv Area', 'Bsmt Full Bath',
       'Bsmt Half Bath', 'Full Bath', 'Half Bath', 'Bedroom AbvGr',
       'Kitchen AbvGr', 'Kitchen Qual', 'TotRms AbvGrd', 'Functional',
       'Fireplaces', 'Fireplace Qu', 'Garage Type', 'Garage Yr Blt',
       'G

In [78]:
train_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2051 entries, 0 to 2050
Data columns (total 81 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Id               2051 non-null   int64  
 1   PID              2051 non-null   int64  
 2   MS SubClass      2051 non-null   int64  
 3   MS Zoning        2051 non-null   object 
 4   Lot Frontage     1721 non-null   float64
 5   Lot Area         2051 non-null   int64  
 6   Street           2051 non-null   object 
 7   Alley            140 non-null    object 
 8   Lot Shape        2051 non-null   object 
 9   Land Contour     2051 non-null   object 
 10  Utilities        2051 non-null   object 
 11  Lot Config       2051 non-null   object 
 12  Land Slope       2051 non-null   object 
 13  Neighborhood     2051 non-null   object 
 14  Condition 1      2051 non-null   object 
 15  Condition 2      2051 non-null   object 
 16  Bldg Type        2051 non-null   object 
 17  House Style   

In [79]:
list_na_train_df = train_df.columns[train_df.isna().any()].to_list()
list_na_train_df

['Lot Frontage',
 'Alley',
 'Mas Vnr Type',
 'Mas Vnr Area',
 'Bsmt Qual',
 'Bsmt Cond',
 'Bsmt Exposure',
 'BsmtFin Type 1',
 'BsmtFin SF 1',
 'BsmtFin Type 2',
 'BsmtFin SF 2',
 'Bsmt Unf SF',
 'Total Bsmt SF',
 'Bsmt Full Bath',
 'Bsmt Half Bath',
 'Fireplace Qu',
 'Garage Type',
 'Garage Yr Blt',
 'Garage Finish',
 'Garage Cars',
 'Garage Area',
 'Garage Qual',
 'Garage Cond',
 'Pool QC',
 'Fence',
 'Misc Feature']

In [80]:
train_df.describe()

Id           PID  MS SubClass  Lot Frontage       Lot Area  \
count  2051.000000  2.051000e+03  2051.000000   1721.000000    2051.000000   
mean   1474.033642  7.135900e+08    57.008776     69.055200   10065.208191   
std     843.980841  1.886918e+08    42.824223     23.260653    6742.488909   
min       1.000000  5.263011e+08    20.000000     21.000000    1300.000000   
25%     753.500000  5.284581e+08    20.000000     58.000000    7500.000000   
50%    1486.000000  5.354532e+08    50.000000     68.000000    9430.000000   
75%    2198.000000  9.071801e+08    70.000000     80.000000   11513.500000   
max    2930.000000  9.241520e+08   190.000000    313.000000  159000.000000   

       Overall Qual  Overall Cond   Year Built  Year Remod/Add  Mas Vnr Area  \
count   2051.000000   2051.000000  2051.000000     2051.000000   2029.000000   
mean       6.112140      5.562165  1971.708922     1984.190151     99.695909   
std        1.426271      1.104497    30.177889       21.036250    174.963129   
min        1.000000      1.000000  1872.000000     1950.000000      0.000000   
25%        5.000000      5.000000  1953.500000     1964.500000      0.000000   
50%        6.000000      5.000000  1974.000000     1993.000000      0.000000   
75%        7.000000      6.000000  2001.000000     2004.000000    161.000000   
max       10.000000      9.000000  2010.000000     2010.000000   1600.000000   

       ...  Wood Deck SF  Open Porch SF  Enclosed Porch   3Ssn Porch  \
count  ...   2051.000000    2051.000000     2051.000000  2051.000000   
mean   ...     93.833740      47.556802       22.571916     2.591419   
std    ...    128.549416      66.747241       59.845110    25.229615   
min    ...      0.000000       0.000000        0.000000     0.000000   
25%    ...      0.000000       0.000000        0.000000     0.000000   
50%    ...      0.000000      27.000000        0.000000     0.000000   
75%    ...    168.000000      70.000000        0.000000     0.000000   
max    ...   1424.000000     547.000000      432.000000   508.000000   

       Screen Porch    Pool Area      Misc Val      Mo Sold      Yr Sold  \
count   2051.000000  2051.000000   2051.000000  2051.000000  2051.000000   
mean      16.511458     2.397855     51.574354     6.219893  2007.775719   
std       57.374204    37.782570    573.393985     2.744736     1.312014   
min        0.000000     0.000000      0.000000     1.000000  2006.000000   
25%        0.000000     0.000000      0.000000     4.000000  2007.000000   
50%        0.000000     0.000000      0.000000     6.000000  2008.000000   
75%        0.000000     0.000000      0.000000     8.000000  2009.000000   
max      490.000000   800.000000  17000.000000    12.000000  2010.000000   

           SalePrice  
count    2051.000000  
mean   181469.701609  
std     79258.659352  
min     12789.000000  
25%    129825.000000  
50%    162500.000000  
75%    214000.000000  
max    611657.000000  

[8 rows x 39 columns]

### "Test" dataset:

In [81]:
test_df.shape

(878, 80)

In [82]:
test_df.head()

Id        PID  MS SubClass MS Zoning  Lot Frontage  Lot Area Street  \
0  2658  902301120          190        RM          69.0      9142   Pave   
1  2718  905108090           90        RL           NaN      9662   Pave   
2  2414  528218130           60        RL          58.0     17104   Pave   
3  1989  902207150           30        RM          60.0      8520   Pave   
4   625  535105100           20        RL           NaN      9500   Pave   

  Alley Lot Shape Land Contour  ... 3Ssn Porch Screen Porch Pool Area Pool QC  \
0  Grvl       Reg          Lvl  ...          0            0         0     NaN   
1   NaN       IR1          Lvl  ...          0            0         0     NaN   
2   NaN       IR1          Lvl  ...          0            0         0     NaN   
3   NaN       Reg          Lvl  ...          0            0         0     NaN   
4   NaN       IR1          Lvl  ...          0          185         0     NaN   

  Fence Misc Feature Misc Val Mo Sold  Yr Sold  Sale Type  
0   NaN          NaN        0       4     2006        WD   
1   NaN          NaN        0       8     2006        WD   
2   NaN          NaN        0       9     2006        New  
3   NaN          NaN        0       7     2007        WD   
4   NaN          NaN        0       7     2009        WD   

[5 rows x 80 columns]

In [83]:
test_df.columns

Index(['Id', 'PID', 'MS SubClass', 'MS Zoning', 'Lot Frontage', 'Lot Area',
       'Street', 'Alley', 'Lot Shape', 'Land Contour', 'Utilities',
       'Lot Config', 'Land Slope', 'Neighborhood', 'Condition 1',
       'Condition 2', 'Bldg Type', 'House Style', 'Overall Qual',
       'Overall Cond', 'Year Built', 'Year Remod/Add', 'Roof Style',
       'Roof Matl', 'Exterior 1st', 'Exterior 2nd', 'Mas Vnr Type',
       'Mas Vnr Area', 'Exter Qual', 'Exter Cond', 'Foundation', 'Bsmt Qual',
       'Bsmt Cond', 'Bsmt Exposure', 'BsmtFin Type 1', 'BsmtFin SF 1',
       'BsmtFin Type 2', 'BsmtFin SF 2', 'Bsmt Unf SF', 'Total Bsmt SF',
       'Heating', 'Heating QC', 'Central Air', 'Electrical', '1st Flr SF',
       '2nd Flr SF', 'Low Qual Fin SF', 'Gr Liv Area', 'Bsmt Full Bath',
       'Bsmt Half Bath', 'Full Bath', 'Half Bath', 'Bedroom AbvGr',
       'Kitchen AbvGr', 'Kitchen Qual', 'TotRms AbvGrd', 'Functional',
       'Fireplaces', 'Fireplace Qu', 'Garage Type', 'Garage Yr Blt',
       'G

In [84]:
test_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 878 entries, 0 to 877
Data columns (total 80 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Id               878 non-null    int64  
 1   PID              878 non-null    int64  
 2   MS SubClass      878 non-null    int64  
 3   MS Zoning        878 non-null    object 
 4   Lot Frontage     718 non-null    float64
 5   Lot Area         878 non-null    int64  
 6   Street           878 non-null    object 
 7   Alley            58 non-null     object 
 8   Lot Shape        878 non-null    object 
 9   Land Contour     878 non-null    object 
 10  Utilities        878 non-null    object 
 11  Lot Config       878 non-null    object 
 12  Land Slope       878 non-null    object 
 13  Neighborhood     878 non-null    object 
 14  Condition 1      878 non-null    object 
 15  Condition 2      878 non-null    object 
 16  Bldg Type        878 non-null    object 
 17  House Style     

In [85]:
list_na_test_df = test_df.columns[test_df.isna().any()].to_list()

# Notes on the EDA (Prior to cleaning):

### Differences between the sets:

- The test contains 1 less column than train: test does not contain "SalePrice"
    - The "SalesPrice" column contains all <b><i>"except for the target that we are trying to predict in our Regression model."</i></b>
- The column names in both sets will need to be reformatted, to make them easier to work with.
    - <b>A function can be used to iterate through and automate necessary changes.</b>
- What are the missing values?
    - How many values should each col contain, including na?
        - <b>Train: 2051 rows
        - Test: 878 rows</b>
    - What cols have missing values?
        - <b>Train: there are 26 cols with NaN
        - Test: there are 19 cols with NaN</b>
    - What do these missing values mean?
        - <b>Missing values in some cols will mean an amenity or feature is not present at that particular parcel (i.e. does not have a pool, or garage, or fireplace, etc).
        - For the BSMT sq foot cols, NaN could be problematic. Does this unit not have a basement?
        - </b>
- Variable types info (available via the "data description":
    - <b>Nominal: Named, categorical
    - Ordinal: named, ranked
    - Discrete: dates, # full/half bathrooms, # bedrooms, # kitchens, # rooms 
    - Continuous: size (sq footage), $ value of misc. feature, SALE PRICE</b>
- Identify Outliers:
    - <b>Approach: (Could use a function for this)
        - Find 50, 25, and 75%'s for dataset
        - Anything < 1.5 * the IQR under Q1, or
        - Anything > 1.5 * the IQR above Q3</b>

# Clean the data:

In [86]:
train_df.columns

Index(['Id', 'PID', 'MS SubClass', 'MS Zoning', 'Lot Frontage', 'Lot Area',
       'Street', 'Alley', 'Lot Shape', 'Land Contour', 'Utilities',
       'Lot Config', 'Land Slope', 'Neighborhood', 'Condition 1',
       'Condition 2', 'Bldg Type', 'House Style', 'Overall Qual',
       'Overall Cond', 'Year Built', 'Year Remod/Add', 'Roof Style',
       'Roof Matl', 'Exterior 1st', 'Exterior 2nd', 'Mas Vnr Type',
       'Mas Vnr Area', 'Exter Qual', 'Exter Cond', 'Foundation', 'Bsmt Qual',
       'Bsmt Cond', 'Bsmt Exposure', 'BsmtFin Type 1', 'BsmtFin SF 1',
       'BsmtFin Type 2', 'BsmtFin SF 2', 'Bsmt Unf SF', 'Total Bsmt SF',
       'Heating', 'Heating QC', 'Central Air', 'Electrical', '1st Flr SF',
       '2nd Flr SF', 'Low Qual Fin SF', 'Gr Liv Area', 'Bsmt Full Bath',
       'Bsmt Half Bath', 'Full Bath', 'Half Bath', 'Bedroom AbvGr',
       'Kitchen AbvGr', 'Kitchen Qual', 'TotRms AbvGrd', 'Functional',
       'Fireplaces', 'Fireplace Qu', 'Garage Type', 'Garage Yr Blt',
       'G

In [99]:
# Reformat col names:

def format_col(df):
    # To lowercase
    df.columns = map(str.lower, df.columns)  # <-- Referenced: https://chrisalbon.com/python/data_wrangling/pandas_lowercase_column_names/
    
    # Remove spaces, add "_":
    df.columns = df.columns.str.replace(" ", "_")
    
    

format_col(train_df)



train_df.rename(columns={"saleprice": "sale_price"}, inplace=True)

train_df.columns



- What cols have missing values?
    - Train: there are 26 cols with NaN
    - Test: there are 19 cols with NaN
- What do these missing values mean?
    - Missing values in some cols will mean an amenity or feature is not present at that particular parcel (i.e. does not have a pool, or garage, or fireplace, etc).
    - For the BSMT sq foot cols, NaN could be problematic. Does this unit not have a basement?
    - 

In [121]:
# NaN's
train_df.isna().sum()

id                0
pid               0
ms_subclass       0
ms_zoning         0
lot_frontage    330
               ... 
misc_val          0
mo_sold           0
yr_sold           0
sale_type         0
sale_price        0
Length: 81, dtype: int64

In [124]:
list_na_train_df = train_df.columns[train_df.isna().any()].to_list()
list_na_train_df

['lot_frontage',
 'alley',
 'mas_vnr_type',
 'mas_vnr_area',
 'bsmt_qual',
 'bsmt_cond',
 'bsmt_exposure',
 'bsmtfin_type_1',
 'bsmtfin_sf_1',
 'bsmtfin_type_2',
 'bsmtfin_sf_2',
 'bsmt_unf_sf',
 'total_bsmt_sf',
 'bsmt_full_bath',
 'bsmt_half_bath',
 'fireplace_qu',
 'garage_type',
 'garage_yr_blt',
 'garage_finish',
 'garage_cars',
 'garage_area',
 'garage_qual',
 'garage_cond',
 'pool_qc',
 'fence',
 'misc_feature']

In [125]:
train_df.loc[train_df["lot_frontage"].isna()]  # <-- No linear feet of street connected to property?

id        pid  ms_subclass ms_zoning  lot_frontage  lot_area street  \
0      109  533352170           60        RL           NaN     13517   Pave   
7      145  535154050           20        RL           NaN     12160   Pave   
8     1942  535353130           20        RL           NaN     15783   Pave   
23      12  527165230           20        RL           NaN      7980   Pave   
27    1534  909277040           50        RL           NaN     11700   Pave   
...    ...        ...          ...       ...           ...       ...    ...   
2016   560  532376110           20        RL           NaN      7791   Pave   
2022  2872  909475020           20        RL           NaN     16381   Pave   
2028  2526  534127190           20        RL           NaN     20781   Pave   
2031    25  527402250           20        RL           NaN     12537   Pave   
2047   785  905377130           30        RL           NaN     12342   Pave   

     alley lot_shape land_contour  ... screen_porch pool_area pool_qc  fence  \
0      NaN       IR1          Lvl  ...            0         0     NaN    NaN   
7      NaN       IR1          Lvl  ...            0         0     NaN  MnPrv   
8      NaN       Reg          Lvl  ...            0         0     NaN  MnPrv   
23     NaN       IR1          Lvl  ...            0         0     NaN  GdPrv   
27    Grvl       IR1          HLS  ...            0         0     NaN    NaN   
...    ...       ...          ...  ...          ...       ...     ...    ...   
2016   NaN       IR1          Lvl  ...            0         0     NaN   GdWo   
2022   NaN       IR1          Lvl  ...            0         0     NaN    NaN   
2028   NaN       IR2          Lvl  ...            0         0     NaN    NaN   
2031   NaN       IR1          Lvl  ...            0         0     NaN    NaN   
2047   NaN       IR1          Lvl  ...            0         0     NaN    NaN   

     misc_feature misc_val mo_sold yr_sold  sale_type  sale_price  
0             NaN        0       3    2010        WD       130500  
7             NaN        0       5    2010        COD      142000  
8            Shed      400       6    2007        WD       112500  
23           Shed      500       3    2010        WD       185000  
27            NaN        0       7    2008        WD       198000  
...           ...      ...     ...     ...        ...         ...  
2016          NaN        0      10    2009        WD       129000  
2022          NaN        0      12    2006        WD       223000  
2028          NaN        0       6    2006        WD       262500  
2031          NaN        0       4    2010        WD       149900  
2047          NaN        0       3    2009        WD        82500  

[330 rows x 81 columns]

In [128]:
train_df.loc[train_df["fence"].isna()]

id        pid  ms_subclass ms_zoning  lot_frontage  lot_area street  \
0      109  533352170           60        RL           NaN     13517   Pave   
1      544  531379050           60        RL          43.0     11492   Pave   
2      153  535304180           20        RL          68.0      7922   Pave   
3      318  916386060           60        RL          73.0      9802   Pave   
4      255  906425045           50        RL          82.0     14235   Pave   
...    ...        ...          ...       ...           ...       ...    ...   
2046  1587  921126030           20        RL          79.0     11449   Pave   
2047   785  905377130           30        RL           NaN     12342   Pave   
2048   916  909253010           50        RL          57.0      7558   Pave   
2049   639  535179160           20        RL          80.0     10400   Pave   
2050    10  527162130           60        RL          60.0      7500   Pave   

     alley lot_shape land_contour  ... screen_porch pool_area pool_qc fence  \
0      NaN       IR1          Lvl  ...            0         0     NaN   NaN   
1      NaN       IR1          Lvl  ...            0         0     NaN   NaN   
2      NaN       Reg          Lvl  ...            0         0     NaN   NaN   
3      NaN       Reg          Lvl  ...            0         0     NaN   NaN   
4      NaN       IR1          Lvl  ...            0         0     NaN   NaN   
...    ...       ...          ...  ...          ...       ...     ...   ...   
2046   NaN       IR1          HLS  ...            0         0     NaN   NaN   
2047   NaN       IR1          Lvl  ...            0         0     NaN   NaN   
2048   NaN       Reg          Bnk  ...            0         0     NaN   NaN   
2049   NaN       Reg          Lvl  ...            0         0     NaN   NaN   
2050   NaN       Reg          Lvl  ...            0         0     NaN   NaN   

     misc_feature misc_val mo_sold yr_sold  sale_type  sale_price  
0             NaN        0       3    2010        WD       130500  
1             NaN        0       4    2009        WD       220000  
2             NaN        0       1    2010        WD       109000  
3             NaN        0       4    2010        WD       174000  
4             NaN        0       3    2010        WD       138500  
...           ...      ...     ...     ...        ...         ...  
2046          NaN        0       1    2008        WD       298751  
2047          NaN        0       3    2009        WD        82500  
2048          NaN        0       3    2009        WD       177000  
2049          NaN        0      11    2009        WD       144000  
2050          NaN        0       6    2010        WD       189000  

[1651 rows x 81 columns]